pkg/reconciler/pipelinerun/controller.go

```diff
@@ -93,11 +97,38 @@ func NewController(images pipeline.Images) func(context.Context, configmap.Watch
 			DeleteFunc: impl.Enqueue,
 		})
 
+		AddPipelineRun := func(obj interface{}) {
+			pr := obj.(*v1alpha1.PipelineRun)
+			c.Logger.Infof("Adding PipelineRun %s/%s", pr.Namespace, pr.Name)
+
+			if pr.DeletionTimestamp == nil && pipelineRunCleanup(pr) {
+				impl.Enqueue(pr)
+			}
+		}
+
+		UpdatePipelineRun := func(old, cur interface{}) {
+			pr := cur.(*v1alpha1.PipelineRun)
+			c.Logger.Infof("Updating PipelineRun %s/%s", pr.Namespace, pr.Name)
+
+			if pr.DeletionTimestamp == nil && pipelineRunCleanup(pr) {
+				impl.Enqueue(pr)
+			}
+		}
+
+		pipelineRunInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+			AddFunc:    AddPipelineRun,
+			UpdateFunc: UpdatePipelineRun,
+		})
```

Comment:
it might look nice to create a new struct that has the UpdatePipelineRun and AddPipelineRun functions in a separate package - then we could have unit tests for them, and all all that would need to happen here would be something like

```go
                expirationQueue := expiration.Queue{}
		pipelineRunInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
			AddFunc:    expirationQueue.AddPipelineRun,
			UpdateFunc: expirationQueue.UpdatePipelineRun,
		})
```

Reply:
Sorry about delaying to resolve this PR. Recently, we have been encapsulating tekton pipeline and dashboard so that we can use tekton in production.

And thank you for the suggestions again, I will seriously consider them!  :)